### PR TITLE
llext export syscalls fixes

### LINKED ISSUE #70582: llext: sem EXPORT_SYMBOL not working in userspace
**Describe the bug**
Discovered on ARM qemu e.g. mps2/an385 while working on #70317, reproduceable with [this branch](https://github.com/laurenmurphyx64/zephyr/commit/5835b2fd276f7a8e0008032b3069ae86e8f8a8f1) and

```
west build -p -b mps2/an385 tests/subsys/llext/simple -- -DCONFIG_USERSPACE=y  && west build -t run
```

EXPORT_SYMBOL for k_thread_join, etc. working, semaphore not

**Expected behavior**
Export succeeds 

**Impact**
Showstopper 

**Logs and console output**
I printed out the contents of the built-in symbol table with and without userspace.

Without userspace:
```
assert_post_action
assert_print
k_is_in_isr
memcmp
memcpy
memset
printk
strcmp
strcpy
strlen
strncmp
strncpy
vprintk
z_impl_k_sem_give
z_impl_k_sem_init
z_impl_k_sem_take
z_impl_k_thread_create
z_impl_k_thread_join
z_impl_k_thread_start
z_log_minimal_printk
ztest_relative_filename
ztest_test_fail
ztest_test_pass
ztest_test_skip
```

With userspace:
```
assert_post_action
assert_print
k_is_in_isr
memcmp
memcpy
memset
printk
strcmp
strcpy
strlen
strncmp
strncpy
vprintk
z_arm_thread_is_in_user_mode
z_impl_k_thread_create
z_impl_k_thread_join
z_impl_k_thread_start
z_log_minimal_printk
z_mrsh___posix_clock_get_base
z_mrsh___ztest_set_test_phase
z_mrsh___ztest_set_test_result
z_mrsh_device_get_binding
z_mrsh_device_is_ready
z_mrsh_gpio_get_pending_int
z_mrsh_gpio_pin_configure
z_mrsh_gpio_pin_get_config
z_mrsh_gpio_pin_interrupt_configure
z_mrsh_gpio_port_clear_bits_raw
z_mrsh_gpio_port_get_direction
z_mrsh_gpio_port_get_raw
z_mrsh_gpio_port_set_bits_raw
z_mrsh_gpio_port_set_masked_raw
z_mrsh_gpio_port_toggle_bits
z_mrsh_k_busy_wait
z_mrsh_k_condvar_broadcast
z_mrsh_k_condvar_init
z_mrsh_k_condvar_signal
z_mrsh_k_condvar_wait
z_mrsh_k_event_clear
z_mrsh_k_event_init
z_mrsh_k_event_post
z_mrsh_k_event_set_masked
z_mrsh_k_event_set
z_mrsh_k_event_wait_all
z_mrsh_k_event_wait
z_mrsh_k_float_disable
z_mrsh_k_float_enable
z_mrsh_k_futex_wait
z_mrsh_k_futex_wake
z_mrsh_k_is_preempt_thread
z_mrsh_k_msgq_alloc_init
z_mrsh_k_msgq_get_attrs
z_mrsh_k_msgq_get
z_mrsh_k_msgq_num_free_get
z_mrsh_k_msgq_num_used_get
z_mrsh_k_msgq_peek_at
z_mrsh_k_msgq_peek
z_mrsh_k_msgq_purge
z_mrsh_k_msgq_put
z_mrsh_k_mutex_init
z_mrsh_k_mutex_lock
z_mrsh_k_mutex_unlock
z_mrsh_k_object_access_grant
z_mrsh_k_object_alloc_size
z_mrsh_k_object_alloc
z_mrsh_k_object_release
z_mrsh_k_pipe_alloc_init
z_mrsh_k_pipe_buffer_flush
z_mrsh_k_pipe_flush
z_mrsh_k_pipe_get
z_mrsh_k_pipe_put
z_mrsh_k_pipe_read_avail
z_mrsh_k_pipe_write_avail
z_mrsh_k_poll_signal_check
z_mrsh_k_poll_signal_init
z_mrsh_k_poll_signal_raise
z_mrsh_k_poll_signal_reset
z_mrsh_k_poll
z_mrsh_k_queue_alloc_append
z_mrsh_k_queue_alloc_prepend
z_mrsh_k_queue_cancel_wait
z_mrsh_k_queue_get
z_mrsh_k_queue_init
z_mrsh_k_queue_is_empty
z_mrsh_k_queue_peek_head
z_mrsh_k_queue_peek_tail
z_mrsh_k_sched_current_thread_query
z_mrsh_k_sem_count_get
z_mrsh_k_sem_give
z_mrsh_k_sem_init
z_mrsh_k_sem_reset
z_mrsh_k_sem_take
z_mrsh_k_sleep
z_mrsh_k_stack_alloc_init
z_mrsh_k_stack_pop
z_mrsh_k_stack_push
z_mrsh_k_str_out
z_mrsh_k_thread_abort
z_mrsh_k_thread_create
z_mrsh_k_thread_custom_data_get
z_mrsh_k_thread_custom_data_set
z_mrsh_k_thread_deadline_set
z_mrsh_k_thread_join
z_mrsh_k_thread_name_copy
z_mrsh_k_thread_name_set
z_mrsh_k_thread_priority_get
z_mrsh_k_thread_priority_set
z_mrsh_k_thread_resume
z_mrsh_k_thread_stack_alloc
z_mrsh_k_thread_stack_free
z_mrsh_k_thread_stack_space_get
z_mrsh_k_thread_start
z_mrsh_k_thread_suspend
z_mrsh_k_thread_timeout_expires_ticks
z_mrsh_k_thread_timeout_remaining_ticks
z_mrsh_k_timer_expires_ticks
z_mrsh_k_timer_remaining_ticks
z_mrsh_k_timer_start
z_mrsh_k_timer_status_get
z_mrsh_k_timer_status_sync
z_mrsh_k_timer_stop
z_mrsh_k_timer_user_data_get
z_mrsh_k_timer_user_data_set
z_mrsh_k_uptime_ticks
z_mrsh_k_usleep
z_mrsh_k_wakeup
z_mrsh_k_yield
z_mrsh_log_buffered_cnt
z_mrsh_log_filter_set
z_mrsh_log_frontend_filter_set
z_mrsh_log_panic
z_mrsh_log_process
z_mrsh_sys_clock_hw_cycles_per_sec_runtime_get
z_mrsh_sys_clock_tick_set
z_mrsh_uart_config_get
z_mrsh_uart_configure
z_mrsh_uart_drv_cmd
z_mrsh_uart_err_check
z_mrsh_uart_irq_err_disable
z_mrsh_uart_irq_err_enable
z_mrsh_uart_irq_is_pending
z_mrsh_uart_irq_rx_disable
z_mrsh_uart_irq_rx_enable
z_mrsh_uart_irq_tx_disable
z_mrsh_uart_irq_tx_enable
z_mrsh_uart_irq_update
z_mrsh_uart_line_ctrl_get
z_mrsh_uart_line_ctrl_set
z_mrsh_uart_poll_in
z_mrsh_uart_poll_in_u16
z_mrsh_uart_poll_out
z_mrsh_uart_poll_out_u16
z_mrsh_uart_rx_disable
z_mrsh_uart_rx_enable
z_mrsh_uart_rx_enable_u16
z_mrsh_uart_tx_abort
z_mrsh_uart_tx
z_mrsh_uart_tx_u16
z_mrsh_z_log_msg_simple_create_0
z_mrsh_z_log_msg_simple_create_1
z_mrsh_z_log_msg_simple_create_2
z_mrsh_z_log_msg_static_create
z_mrsh_z_sys_mutex_kernel_lock
z_mrsh_z_sys_mutex_kernel_unlock
z_mrsh_z_test_1cpu_start
z_mrsh_z_test_1cpu_stop
z_mrsh_zephyr_fputc
z_mrsh_zephyr_fwrite
z_mrsh_zephyr_read_stdin
z_mrsh_zephyr_write_stdout
z_mrsh_ztest_run_test_suites
z_mrsh_ztest_set_assert_valid
z_mrsh_ztest_set_fault_valid
ztest_relative_filename
ztest_test_fail
ztest_test_pass
ztest_test_skip
```

**Environment (please complete the following information):**
 - OS: Ubuntu 22
 - Toolchain: Zephyr SDK
 - `3b36911`

### PR DESCRIPTION
**Pull request by edersondisouza**
With 67bb6db3f8e5329516f005f1335bb7a23a95b3eb, one doesn't need to export manually each syscall available. However, there are some flaws there.

First is that it only works if the extension lives on userspace. It was exporting the `z_mrsh_` symbols, available to userspace. If an extension actually lives in kernelspace, it wouldn't have the `z_impl_` symbol available.

Second flaw is that syscalls could still be optimised away if the Zephyr application itself wasn't using them. This is the issue of #70582 - and it was there before 67bb6db3f8e5329516f005f1335bb7a23a95b3eb.

This PR fixes both flaws: the first by exporting `z_impl_` symbols instead of `z_mrsh_` ones. The second, by including `libkernel.a` into the `whole-archive` during linking. Unfortunately both changes are a bit more invasive than the original patch. It also adds tests for using syscalls from extensions.

(If you feel like 67bb6db3f8e5329516f005f1335bb7a23a95b3eb was some sort of trojan horse to get these new changes, I know that feel and I'm sorry - I ended up not testing 67bb6db3f8e5329516f005f1335bb7a23a95b3eb as thoroughly as I should, otherwise I'd caught these flaws before sending it. I feel like I wasted reviewers time... =/ )

Fixes #70582

### CODE REVIEW

File: scripts/build/gen_syscalls.py [hunk 1]
```diff
@@ -159,6 +157,20 @@
 #endif
 """
 
+
+exported_template = """
+/* Export syscalls for extensions */
+static void *no_handler = NULL;
```
Comment (pillo79) [inline scripts/build/gen_syscalls.py:163]:
```suggestion
static void * const no_handler = NULL;
```
Should spare a bit of RAM, but most importantly, hopefully relocs :slightly_smiling_face:

**Review by pillo79 — DISMISSED** [review]
LGTM, a few suggestions if you wish to consider them.

**Comment by edersondisouza** [comment]
v2:
  - More consistent naming for cmake variable;
  - More `const`s;
  - Rebased.

File: cmake/linker/arcmwdt/target.cmake [hunk 2]
```diff
@@ -97,9 +97,9 @@ function(toolchain_ld_link_elf)
     ${LINKERFLAGPREFIX}--entry=__start
     ${LINKERFLAGPREFIX}--Map=${TOOLCHAIN_LD_LINK_ELF_OUTPUT_MAP}
     ${LINKERFLAGPREFIX}--whole-archive
-    ${ZEPHYR_LIBS_PROPERTY}
-    ${LINKERFLAGPREFIX}--no-whole-archive
-    kernel
+    ${WHOLE_ARCHIVE_LIBS}
+    ${LINKERFLAGPREFIX},--no-whole-archive
```
Comment (pillo79) [inline cmake/linker/arcmwdt/target.cmake:101]: Tthis linker, unlike all others, does not seem to use commas... :man_shrugging: is this a typo?

**Review by pillo79 — COMMENTED** [review]
One tiny bug in v2?

File: cmake/linker/arcmwdt/target.cmake (hunk 2 above)
Comment (edersondisouza) [reply cmake/linker/arcmwdt/target.cmake:101]: Oh noes, copypasta detected!

**Comment by edersondisouza** [comment]
v3:
 - Fix typo.

**Review by teburd — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]

**Comment by teburd** [comment]
@tejlmand arguably this is an llext related PR and so I've reassigned to myself, if this feels wrong please do change

**Review by stephanosio — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**